src/message/MoveMessage.js

```diff
@@ -0,0 +1,216 @@
+import TopicAutocomplete from '../autocomplete/TopicAutocomplete';
+import { streamNameOfNarrow, streamNarrow } from '../utils/narrow';
+
+type Props = $ReadOnly<{|
+  navigation: AppNavigationProp<'move-message'>,
+  route: RouteProp<'move-message', {| message: Message | Outbox, messageNarrow: Narrow |}>,
+|}>;
+
+const inputMarginPadding = {
+  paddingHorizontal: 8,
+  paddingVertical: Platform.select({
+    ios: 8,
+    android: 2,
+  }),
+};
+
+export default function MoveMessage(props: Props): Node {
+  const themeContext = useContext(ThemeContext);
+  const backgroundColor = themeContext.backgroundColor;
+  const cardColor = themeContext.cardColor;
+  const iconName = Platform.OS === 'android' ? 'arrow-left' : 'chevron-left';
+  const auth = useSelector(getAuth);
+  const allStreams = useSelector(getStreams);
+  const isAdmin = useSelector(getOwnUser).is_admin;
+  const messageId = props.route.params.message.id;
+  const currentStreamName = streamNameOfNarrow(props.route.params.messageNarrow);
+  const currentStreamId = allStreams.find(s => s.name === currentStreamName)?.stream_id;
+  const [narrow, setNarrow] = useState(streamNarrow(currentStreamName));
+  const [subject, setSubject] = useState(props.route.params.message.subject);
+  const [propagateMode, setPropagateMode] = useState('change_one');
+  const [streamId, setStreamId] = useState(currentStreamId);
+  const [topicFocus, setTopicFocus] = useState(false);
+
+  const styles = {
+    parent: {
+      backgroundColor: cardColor,
+    },
+    layout: {
+      margin: 10,
+    },
+    title: {
+      fontSize: 18,
+      color: backgroundColor === 'white' ? 'black' : 'white',
+    },
+    topicInput: {
+      height: 50,
+      backgroundColor,
+      ...inputMarginPadding,
+    },
+    viewTitle: {
+      display: 'flex',
+      flexDirection: 'row',
+      justifyContent: 'space-between',
+      alignItems: 'center',
+      height: 50,
+      paddingHorizontal: 10,
+      marginBottom: 20,
+    },
+    textColor: {
+      color: backgroundColor === 'white' ? 'black' : 'white',
+    },
+    picker: { backgroundColor, marginBottom: 20 },
+    submitButton: {
+      marginTop: 10,
+      paddingTop: 15,
+      paddingBottom: 15,
+      marginLeft: 30,
+      marginRight: 30,
+      backgroundColor: BRAND_COLOR,
+      borderRadius: 10,
+      borderWidth: 1,
+    },
+  };
+
+  const handleTopicChange = (topic: string) => {
+    setSubject(topic);
+  };
+
+  const handleTopicFocus = () => {
+    setTopicFocus(true);
+  };
+
+  const setTopicInputValue = (topic: string) => {
+    handleTopicChange(topic);
+    setTopicFocus(false);
+  };
+
+  const handleTopicAutocomplete = (topic: string) => {
+    setTopicInputValue(topic);
+  };
+
+  const updateMessage = () => {
+    if (isAdmin) {
+      api
+        .updateMessage(
+          auth,
+          { subject, stream_id: streamId, propagate_mode: propagateMode },
+          messageId,
+        )
+        .catch(error => {
```

Comment:
Don't use `Promise#catch`; instead, make the surrounding function async, and use `await` inside a try/catch block. (You can search for `await api` to find a number of examples.)

The reason is that with `Promise#catch` it's hard to avoid having subtle edge cases where exceptions don't get handled properly, and using try/catch blocks instead gives a much more robust structure for managing our exception handling.

(This isn't something that's obvious a priori. I wish I had a good writeup of it to link you to, but I don't.)

Reply:
Oh, and here's a related benefit of using try/catch: you have these two `.catch` calls that are completely identical, and it would be better to say that just once. With try/catch, that's very straightforward to do: just put the try/catch outside the if/else that controls the two different paths.